File: leetcode/c++/LCR_131_jian_sheng_zi_lcof.cpp
// https://leetcode.cn/problems/jian-sheng-zi-lcof

#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int cuttingBamboo(int bamboo_len) {
        if (bamboo_len <= 3)
            return bamboo_len - 1;
        int a = bamboo_len / 3, b = bamboo_len % 3;
        if (b == 0)
            return pow(3, a);
        if (b == 2)
            return pow(3, a) * 2;
        return pow(3, a-1) * 4;
    }
};

// 数论，每个数字都能分为2和3（奇偶），3越多越大，比如3*3 > 2*2*2
// 当余数为1时，需要把前一个3和这个1变为2和2，2*2 > 1*3
// O(1)，O(1)
// 也可以用动态规划

File: leetcode/golang/LCR_131_jian_sheng_zi_lcof.go
func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}

func cuttingBamboo(bamboo_len int) int {
    dp := make([]int, bamboo_len+1)
    dp[1], dp[2], dp[3] = 1, 1, 2
    for i := 4; i <= bamboo_len; i++ {
        for j := 1; j < i; j++ {
            dp[i] = max(max(dp[i], j * dp[i-j]), (i-j)*j)
        }
    }
    return dp[bamboo_len]
}